packages/react-devtools-shared/src/backend/renderer.js

```diff
@@ -522,13 +524,129 @@ export function attach(
     typeof setSuspenseHandler === 'function' &&
     typeof scheduleUpdate === 'function';
 
+  // Mapping of fiber IDs to error/warning messages and counts.
+  const fiberToErrorsMap: Map<number, Map<string, number>> = new Map();
+  const fiberToWarningsMap: Map<number, Map<string, number>> = new Map();
+
+  function recordErrorOrWarningOnFiber(
+    type: 'error' | 'warn',
+    args: $ReadOnlyArray<any>,
+    fiberID: number,
+  ): void {
+    const message = format(...args);
+
+    const fiberMap = type === 'error' ? fiberToErrorsMap : fiberToWarningsMap;
+    const messageMap = fiberMap.get(fiberID);
+    if (messageMap != null) {
+      const count = messageMap.get(message) || 0;
+      messageMap.set(message, count + 1);
+    } else {
+      fiberMap.set(fiberID, new Map([[message, 1]]));
+    }
+  }
+
+  function clearErrorsAndWarnings() {
+    const updatedFiberIDs = new Set(fiberToErrorsMap.keys());
+    // eslint-disable-next-line no-for-of-loops/no-for-of-loops
+    for (const key of fiberToWarningsMap.keys()) {
+      updatedFiberIDs.add(key);
+    }
+
+    fiberToErrorsMap.clear();
+    fiberToWarningsMap.clear();
+
+    emitErrorOrWarningUpdates(updatedFiberIDs);
+  }
+
+  function clearErrorsForFiberID(id: number) {
+    if (fiberToErrorsMap.has(id)) {
+      fiberToErrorsMap.delete(id);
+      emitErrorOrWarningUpdates(new Set([id]));
+    }
+  }
+
+  function clearWarningsForFiberID(id: number) {
+    if (fiberToWarningsMap.has(id)) {
+      fiberToWarningsMap.delete(id);
+      emitErrorOrWarningUpdates(new Set([id]));
+    }
+  }
+
+  function onErrorsAndWarningsUpdate(
+    errorsAndWarnings: Array<{|
+      fiber: Fiber,
+      type: 'error' | 'warn',
+      args: $ReadOnlyArray<any>,
+    |}>,
+  ): void {
+    const updatedFiberIDs: Set<number> = new Set();
+    errorsAndWarnings.forEach(({args, fiber, type}) => {
+      if (isFiberMounted(fiber)) {
+        const fiberID = getFiberID(getPrimaryFiber(fiber));
+        updatedFiberIDs.add(fiberID);
+        recordErrorOrWarningOnFiber(type, args, fiberID);
+      }
+    });
+
+    emitErrorOrWarningUpdates(updatedFiberIDs);
+  }
+
+  function emitErrorOrWarningUpdates(updatedIDs: Set<number>) {
+    // TODO (inline errors) Do we just want to push to pendingEvents and flush?
```

Comment:
Regarding this comment, here's kind of what I was thinking:

1. Most errors will probably occur during render– (some will happen during layout and passive effects, but probably fewer). All errors that happen during render _and_ layout will have already happened by the time the DevTools hook gets called, so we can just flush them as part of `flushPendingEvents`.
2. The only errors that won't be covered in that are ones that happen during passive effects. If we waited on `handleCommitFiberRoot` to flush pending events, we'd be waiting on the next commit (that might not even happen) so there's some value in an additional fallback timer like we have that says "ok I give up, I'm going to send them now by themselves".

I think the way we're (now) using a timer in `backend/console` to delay even _telling_ `backend/renderer` about pending render and layout errors forces them to be delayed more than necessary (and also forces an extra render on the front end since it will get the information via separate bridge events). This may be good enough reason to move the timeout _out_ of the `backend/console`.

Now the behavior would be:
1. Queue pending errors and warnings in an "pending" array like we do with pending operations etc.
1. Also set a timer for the (uncommon) case that error was from a passive effect.
1. Check the pending errors/warnings and send them along at the end of `flushPendingEvents` (and clear the timeout).
1. If we actually _do_ hit our timeout, call `flushPendingEvents` again and let it send _only_ those pending events.

Does this make sense? Would you prefer me to make the change or would you like to?

Reply:
Makes all sense. Sounds like you already have a good picture how this might look so it's better if you go ahead and implement it.

I'm currently tracking down https://github.com/facebook/react/pull/20463#issuecomment-746921601. It looks like I'm hitting https://github.com/facebook/react/blob/ac2e861fbe05901b874e3ab49807abab820ef648/packages%2Freact-devtools-extensions%2Fsrc%2FcontentScript.js#L71-L79.

Reply:
Okay. I'll pick up this change then. Good luck with yours!

Reply:
Going to use "reload + profile" from the profiler tab. Then devtools will be setup before the initial mount. Since this happens on master as well I'm not going to dig into further. Maybe this is more common the more extensions one has? We'll see if people report this if we release this PR and rely more on injection timing.